downstairs/src/lib.rs

```diff
@@ -1226,14 +1241,46 @@ where
                         gen,
                         read_only,
                         encrypted,
+                        supported_versions,
                     }) => {
                         if negotiated != 0 {
                             bail!("Received connect out of order {}",
                                 negotiated);
                         }
-
-                        if version != 1 {
-                            bail!("expected version 1, got {}", version);
+                        info!(log, "Connection request from {} with version {}",
+                            upstairs_id, version);
+
+                        // Verify we can communicate with the upstairs.  First
+                        // check our message version.  If that fails,  check
+                        // to see if our version is one of the supported
+                        // versions the upstairs has told us it can support.
+                        if version != CRUCIBLE_MESSAGE_VERSION {
+                            if supported_versions
+                                .contains(&CRUCIBLE_MESSAGE_VERSION)
+                            {
+                                warn!(
+                                    log,
+                                    "downstairs requires down rev version {}",
```

Comment:
with the way this is written now, it doesn't seem like this will always be a down rev? (assuming that means a less recent version)

Reply:
Technically, no, the versions in the "supported_versions" Vec are not required to be lower
than the `CRUCIBLE_MESSAGE_VERSION`.

I thought of this like `CRUCIBLE_MESSAGE_VERSION` is what the `Message` enum would
be versioned as.  If the upstairs in some future decides there were other versions of the
`Message` enum it could support, it would include them in that `supported_versions` Vec.
Like, `CRUCIBLE_MESSAGE_VERSION` is 6, but we can talk to a downstairs with 5 or 4.

None of that code is written yet though.  And, sure, we might send a version that is
higher than `CRUCIBLE_MESSAGE_VERSION`.

Would this read better as `downstairs requires alternate message version {}`?

Reply:
Hm, I don't think that quite feels right. IIUC, the downstairs is reporting that it and the upstairs are at different (but compatible) versions, and that we are carrying on as usual. (I think the "requires" phrasing is throwing me off.) If that's the case, then maybe something like: "downstairs and upstairs using different but compatible versions"? And then I would probably put the version of each in the log message, as well as even the supported upstairs versions.

Reply:
Yes, I'll update to that suggestion, thanks.

Reply:
See if recent changes are to your liking.